fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/ConcreteAvailability.java

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 Yahoo Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.metadata.DataSourceMetadataService;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import org.joda.time.Interval;
+
+import avro.shaded.com.google.common.collect.ImmutableSet;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An availability that provides column and table available interval services for concrete physical table.
+ */
+public class ConcreteAvailability implements Availability {
+
+    private final TableName name;
+    private final Set<Column> columns;
+    private final DataSourceMetadataService metadataService;
+
+    private final Set<String> columnNames;
+
+    /**
+     * Constructor.
+     *
+     * @param tableName  The name of the table and data source associated with this Availability
+     * @param columns  The columns associated with the table and availability
+     * @param metadataService  A service containing the datasource segment data
+     */
+    public ConcreteAvailability(
+            TableName tableName,
+            @NotNull Set<Column> columns,
+            @NotNull DataSourceMetadataService metadataService
+    ) {
+        this.name = tableName;
+        this.columns = ImmutableSet.copyOf(columns);
+        this.metadataService = metadataService;
+
+        this.columnNames = columns.stream().map(Column::getName).collect(Collectors.toSet());
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return Collections.singleton(name);
```

Comment:
I argue that doing it here makes logic clear and easier to maintain, might lose maintainability by trying to do a tiny optimization that might not be needed.